server/src/main/java/org/opensearch/cluster/metadata/ResourceLimitGroup.java

```diff
@@ -0,0 +1,273 @@
+        this(in.readString(), in.readString(), ResourceLimitGroupMode.fromName(in.readString()), in.readMap());
+    }
+
+    /**
+     * Write this into the {@linkplain StreamOutput}.
+     *
+     * @param out
+     */
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(name);
+        out.writeString(_id);
+        out.writeString(mode.getName());
+        out.writeMap(resourceLimits);
+    }
+
+    private void validateResourceLimits(Map<String, Object> resourceLimits) {
+        for (Map.Entry<String, Object> resource : resourceLimits.entrySet()) {
+            String resourceName = resource.getKey();
+            Double threshold = (Double) resource.getValue();
+            Objects.requireNonNull(resourceName, "resourceName can't be null");
+            Objects.requireNonNull(threshold, "resource value can't be null");
+
+            if (Double.compare(threshold, 1.0) > 0) {
+                throw new IllegalArgumentException("resource value should be less than 1.0");
+            }
+
+            if (!ALLOWED_RESOURCES.contains(resourceName.toLowerCase(Locale.ROOT))) {
+                throw new IllegalArgumentException(
+                    "resource has to be valid, valid resources " + ALLOWED_RESOURCES.stream().reduce((x, e) -> x + ", " + e).get()
+                );
+            }
+        }
+    }
+
+    /**
+     * @param builder
+     * @param params
+     * @return
+     * @throws IOException
+     */
+    @Override
+    public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+        builder.startObject();
+        builder.startObject(this.name);
+        builder.field("_id", _id);
+        builder.field("mode", mode.getName());
+        builder.mapContents(resourceLimits);
+        builder.endObject();
+        builder.endObject();
+        return builder;
+    }
+
+    public static ResourceLimitGroup fromXContent(final XContentParser parser) throws IOException {
+        if (parser.currentToken() == null) { // fresh parser? move to the first token
+            parser.nextToken();
+        }
+        if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+            parser.nextToken(); // move to field name
+        }
+        if (parser.currentToken() != XContentParser.Token.FIELD_NAME) {  // on a start object move to next token
+            throw new IllegalArgumentException("Expected FIELD_NAME token but found [" + parser.currentName() + "]");
+        }
+
+        Builder builder = builder().name(parser.currentName());
+
+        XContentParser.Token token = parser.nextToken();
+
+        if (token != XContentParser.Token.START_OBJECT) {
+            throw new IllegalArgumentException("Expected START_OBJECT token but found [" + parser.currentName() + "]");
+        }
+
+        String fieldName = "";
+        // Map to hold resources
+        final Map<String, Object> resourceLimitGroup_ = new HashMap<>();
```

Comment:
Curious why this naming with "_" suffix here, is this some sort of convention?

Reply:
No since this is local to the method, I went ahead with this.